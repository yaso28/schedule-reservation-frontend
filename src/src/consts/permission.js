export default Object.freeze({
  reservation: Object.freeze({
    read: 'reservation.read',
    write: 'reservation.write'
  })
})
