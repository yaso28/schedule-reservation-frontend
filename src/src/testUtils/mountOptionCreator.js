import { createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import routes from '@/router/routes'
import Vuex from 'vuex'
import VueI18n from 'vue-i18n'
import i18nOptions from '@/i18n/options'
import VueMeta from 'vue-meta'
import CoreuiVue from '@coreui/vue'
import icons from '@/icons'
import services from '@/services'

class MountOptionCreator {
  create (storeModules = null) {
    const localVue = createLocalVue()

    // router
    localVue.use(VueRouter)
    const router = new VueRouter({
      mode: 'history',
      base: process.env.BASE_URL,
      routes
    })

    // i18n
    localVue.use(VueI18n)
    const i18n = new VueI18n(i18nOptions)

    // vue-meta
    localVue.use(VueMeta)

    // Core UI
    localVue.use(CoreuiVue)

    const option = {
      localVue,
      router,
      i18n,
      icons,
      provide: services
    }

    // store
    if (storeModules) {
      localVue.use(Vuex)
      option.store = new Vuex.Store({
        modules: storeModules
      })
    }

    return option
  }
}

export default new MountOptionCreator()
