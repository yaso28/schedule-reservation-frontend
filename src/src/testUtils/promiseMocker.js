import _ from 'lodash'

const createSetTimeoutPromise = getResult => new Promise((resolve, reject) => {
  setTimeout(() => getResult(resolve, reject), 1000)
})
const createSwitchPromise = (value, error, switchObj) => createSetTimeoutPromise(
  (resolve, reject) => switchObj.isSuccess ? resolve(value) : reject(error)
)
const createResolvedPromise = value => createSetTimeoutPromise(resolve => resolve(value))
const createRejectedPromise = error => createSetTimeoutPromise((resolve, reject) => reject(error))

const createMock = createPromise => jest.fn().mockImplementationOnce(createPromise)
const createMockForCases = (caseList, createPromiseForCurrentCase) => jest.fn((...args) => {
  for (let i = 0; i < caseList.length; i++) {
    const currentCase = caseList[i]
    if (_.isEqual(args, currentCase.args)) {
      return createPromiseForCurrentCase(currentCase)
    }
  }
  throw new Error('createMockForCases: no args matched')
})

class PromiseMocker {
  init () {
    jest.useFakeTimers()
  }

  clear () {
    jest.clearAllTimers()
  }

  flushPendings () {
    jest.runOnlyPendingTimers()
  }

  createSwitch () {
    return { isSuccess: true }
  }

  setSwitch (switchObj, isSuccess) {
    switchObj.isSuccess = isSuccess
  }

  mock (value, error, switchObj) {
    return createMock(() => createSwitchPromise(value, error, switchObj))
  }

  mockResolved (value) {
    return createMock(() => createResolvedPromise(value))
  }

  mockRejected (error) {
    return createMock(() => createRejectedPromise(error))
  }

  mockForCases (caseList) {
    return createMockForCases(caseList, currentCase => createSwitchPromise(currentCase.value, currentCase.error, currentCase.switch))
  }

  mockResolvedForCases (caseList) {
    return createMockForCases(caseList, currentCase => createResolvedPromise(currentCase.value))
  }

  mockRejectedForCases (caseList) {
    return createMockForCases(caseList, currentCase => createRejectedPromise(currentCase.error))
  }
}

export default new PromiseMocker()
