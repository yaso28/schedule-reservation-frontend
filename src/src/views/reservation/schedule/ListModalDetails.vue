<template>
  <CModal
    :show="visible"
    :title="$t('common.details')"
    color="info"
    centered
    @update:show="updateShow"
  >
    <dl
      v-if="schedule"
      class="row"
    >
      <dt class="col-sm-4">
        {{ $t('schedule.ymd') }}
      </dt>
      <dd class="col-sm-8">
        <MyDateSpan :value="schedule.ymd" />
      </dd>
      <dt class="col-sm-4">
        {{ $t('schedule.begins_at') }}
      </dt>
      <dd class="col-sm-8">
        {{ schedule.begins_at }}
      </dd>
      <dt class="col-sm-4">
        {{ $t('schedule.ends_at') }}
      </dt>
      <dd class="col-sm-8">
        {{ schedule.ends_at }}
      </dd>
      <dt class="col-sm-4">
        {{ $t('schedule_place.model') }}
      </dt>
      <dd class="col-sm-8">
        {{ schedule.schedule_place.name }}
      </dd>
      <dt class="col-sm-4">
        {{ $t('schedule_usage.model') }}
      </dt>
      <dd class="col-sm-8">
        {{ schedule.schedule_usage.name }}
      </dd>
      <dt class="col-sm-4">
        {{ $t('schedule_timetable.model') }}
      </dt>
      <dd
        class="col-sm-8 nl2br"
        v-text="timetable"
      />
      <dt class="col-sm-4">
        {{ $t('reservation_status.model') }}
      </dt>
      <dd class="col-sm-8">
        {{ schedule.reservation_status.name }}
      </dd>
      <dt class="col-sm-4">
        {{ $t('schedule_status.model') }}
      </dt>
      <dd class="col-sm-8">
        {{ schedule.schedule_status.name }}
      </dd>
    </dl>

    <p class="h5">
      {{ $t('reservation.model') }}
    </p>
    <CDataTable
      v-if="reservationList"
      :items="reservationList"
      :fields="fields"
      :header="false"
      add-table-classes="mb-0"
    >
      <template #reservation_status="{item}">
        <td>
          {{ item.reservation_status.name }}
        </td>
      </template>
      <template #action="{item}">
        <td class="my-action">
          <CLink
            v-if="canWrite"
            class="btn btn-outline-primary"
            :to="{ name: 'reservation.reservation.split', params: { id: item.id } }"
          >
            {{ $t('action.split') }}
          </CLink>
        </td>
      </template>
    </CDataTable>

    <template #footer>
      <template v-if="canWrite && schedule">
        <CLink
          class="btn btn-outline-primary"
          :to="{ name: 'reservation.schedule.edit', params: { id: schedule.id } }"
        >
          {{ $t('action.edit') }}
        </CLink>
      </template>
      <template v-else>
          &nbsp;
      </template>
    </template>
  </CModal>
</template>

<script>
import permission from '@/consts/permission'

export default {
  name: 'ListModalDetails',
  components: {
    MyDateSpan: require('@/components/MyDateSpan').default
  },
  inheritAttrs: false,
  inject: ['apiService'],
  props: {
    show: {
      type: Boolean,
      default: false
    },
    schedule: {
      type: Object,
      default: () => null
    }
  },
  data () {
    return {
      visible: this.show,
      reservationList: null
    }
  },
  computed: {
    fields () {
      return [
        { key: 'begins_at', label: this.$t('schedule.begins_at') },
        { key: 'ends_at', label: this.$t('schedule.ends_at') },
        { key: 'reservation_status', label: this.$t('reservation_status.model') },
        { key: 'action', label: '' }
      ]
    },
    timetable () {
      return this.schedule.schedule_timetable?.details ?? ' '
    },
    canWrite () {
      return this.$store.getters['auth/can'][permission.reservation.write]
    }
  },
  watch: {
    show (to) {
      if (to) {
        this.onShow()
      } else {
        this.onHide()
      }
    }
  },
  mounted () {
    if (this.show) {
      this.onShow()
    }
  },
  methods: {
    updateShow (value) {
      this.$emit('update:show', value)
    },
    async onShow () {
      if (!this.schedule) {
        this.updateShow(false)
        return
      }
      try {
        this.$store.commit('alert/clear')
        this.$store.commit('processing/beginGet')
        this.reservationList = await this.apiService.getReservationListForSchedule(this.schedule.id)
        this.visible = true
      } catch (e) {
        this.updateShow(false)
        if (!this.apiService.isHandledError(e)) {
          throw e
        }
      } finally {
        this.$store.commit('processing/endGet')
      }
    },
    onHide () {
      this.visible = false
    }
  }
}
</script>
