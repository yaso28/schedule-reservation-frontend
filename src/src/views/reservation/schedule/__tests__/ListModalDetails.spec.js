import ListModalDetails from '../ListModalDetails'
import mountOptionCreator from '@/testUtils/mountOptionCreator'
import promiseMocker from '@/testUtils/promiseMocker'
import storeModules from '@/store/modules'
import { mount } from '@vue/test-utils'
import permission from '@/consts/permission'

jest.mock('@/store/modules')

const schedule = require('@/testUtils/dummyApiData/schedule/get.json')
const reservationList = require('@/testUtils/dummyApiData/reservation/list-for-schedule-split.json')

const assertClose = function (wrapper) {
  expect(wrapper.emitted()['update:show'][0][0]).toBe(false)
}
const assertNotClose = function (wrapper) {
  expect(wrapper.emitted()['update:show']).toBeFalsy()
}

describe('ListModalDetails', () => {
  let mountOption
  let apiService
  beforeEach(() => {
    promiseMocker.init()
    mountOption = mountOptionCreator.create(storeModules)
    mountOption.propsData = {
      show: false,
      schedule: null
    }
    apiService = mountOption.provide.apiService
  })

  afterEach(() => {
    promiseMocker.clear()
    jest.resetAllMocks()
  })

  describe('write権限：あり', () => {
    beforeEach(() => {
      storeModules.auth.getters.can.mockReturnValue({ [permission.reservation.write]: true })
    })

    describe('マウント', () => {
      let wrapper
      beforeEach(() => {
        wrapper = mount(ListModalDetails, mountOption)
      })

      describe('schedule：なし', () => {
        describe('表示', () => {
          beforeEach(async () => {
            apiService.getReservationListForSchedule = jest.fn()
            await wrapper.setProps({ show: true })
          })

          test('閉じる', () => {
            assertClose(wrapper)
          })

          test('予約データ取得処理は呼び出さない', () => {
            expect(apiService.getReservationListForSchedule).not.toHaveBeenCalled()
          })
        })
      })

      describe('schedule：あり', () => {
        beforeEach(async () => {
          await wrapper.setProps({ schedule })
        })

        describe('表示', () => {
          let loadError
          let loadSwitch
          beforeEach(async () => {
            loadError = new Error('load')
            loadSwitch = promiseMocker.createSwitch()
            apiService.getReservationListForSchedule = promiseMocker.mock(reservationList, loadError, loadSwitch)
            await wrapper.setProps({ show: true })
          })

          test('閉じない', () => {
            assertNotClose(wrapper)
          })

          test('アラートクリア', () => {
            expect(storeModules.alert.mutations.clear).toHaveBeenCalled()
          })

          test('Processingカバー表示開始', () => {
            expect(storeModules.processing.mutations.beginGet).toHaveBeenCalled()
          })

          test('予約データ取得処理を呼出', () => {
            expect(apiService.getReservationListForSchedule).toHaveBeenCalledWith(schedule.id)
          })

          test('Processingカバー表示中', () => {
            expect(storeModules.processing.mutations.endGet).not.toHaveBeenCalled()
          })

          describe('データ取得：エラー', () => {
            beforeEach(() => {
              promiseMocker.setSwitch(loadSwitch, false)
              apiService.isHandledError = jest.fn().mockReturnValueOnce(true)
              promiseMocker.flushPendings()
            })

            test('閉じる', () => {
              assertClose(wrapper)
            })

            test('エラー処理', () => {
              expect(apiService.isHandledError).toHaveBeenCalledWith(loadError)
            })

            test('Processingカバー表示終了', () => {
              expect(storeModules.processing.mutations.endGet).toHaveBeenCalled()
            })
          })

          describe('データ取得：成功', () => {
            beforeEach(() => {
              promiseMocker.flushPendings()
            })

            test('閉じない', () => {
              assertNotClose(wrapper)
            })

            test('モーダル全体表示', () => {
              expect(wrapper.element).toMatchSnapshot()
            })

            test('Processingカバー表示終了', () => {
              expect(storeModules.processing.mutations.endGet).toHaveBeenCalled()
            })
          })
        })
      })
    })
  })

  describe('write権限：なし', () => {
    beforeEach(() => {
      storeModules.auth.getters.can.mockReturnValue({})
    })

    describe('マウント', () => {
      let wrapper
      beforeEach(() => {
        wrapper = mount(ListModalDetails, mountOption)
      })

      describe('表示', () => {
        beforeEach(async () => {
          await wrapper.setProps({ schedule })
          apiService.getReservationListForSchedule = promiseMocker.mockResolved(reservationList)
          await wrapper.setProps({ show: true })
          promiseMocker.flushPendings()
        })

        test('予約一覧アクション表示', () => {
          expect(wrapper.find('td.my-action').element).toMatchSnapshot()
        })

        test('フッター表示', () => {
          expect(wrapper.find('footer.modal-footer').element).toMatchSnapshot()
        })
      })
    })
  })
})
