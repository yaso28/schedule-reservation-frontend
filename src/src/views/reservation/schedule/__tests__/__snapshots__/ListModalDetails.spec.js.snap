// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListModalDetails write権限：あり マウント schedule：あり 表示 データ取得：成功 モーダル全体表示 1`] = `
<div>
  <div
    class="modal overflow-auto fade d-block modal-info"
    role="dialog"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-dialog-centered"
      role="document"
    >
      <div
        class="modal-content"
      >
        <header
          class="modal-header"
        >
          <h5
            class="modal-title"
          >
             詳細 
          </h5>
          <button
            aria-label="Close"
            class="close"
            type="button"
          >
            ×
          </button>
        </header>
        <div
          class="modal-body"
        >
          <dl
            class="row"
          >
            <dt
              class="col-sm-4"
            >
              
      日付
    
            </dt>
             
            <dd
              class="col-sm-8"
            >
              <span>
                
  2020/11/21
  
                <span
                  class="text-primary"
                >
                  
    (土)
  
                </span>
              </span>
            </dd>
             
            <dt
              class="col-sm-4"
            >
              
      開始時刻
    
            </dt>
             
            <dd
              class="col-sm-8"
            >
              
      10:00
    
            </dd>
             
            <dt
              class="col-sm-4"
            >
              
      終了時刻
    
            </dt>
             
            <dd
              class="col-sm-8"
            >
              
      13:00
    
            </dd>
             
            <dt
              class="col-sm-4"
            >
              
      場所
    
            </dt>
             
            <dd
              class="col-sm-8"
            >
              
      体育館＠公民館別棟
    
            </dd>
             
            <dt
              class="col-sm-4"
            >
              
      用途
    
            </dt>
             
            <dd
              class="col-sm-8"
            >
              
      初心者練習
    
            </dd>
             
            <dt
              class="col-sm-4"
            >
              
      時間割
    
            </dt>
             
            <dd
              class="col-sm-8 nl2br"
            >
              1回目：10:00-
2回目：11:30-
            </dd>
             
            <dt
              class="col-sm-4"
            >
              
      予約ステータス
    
            </dt>
             
            <dd
              class="col-sm-8"
            >
              
      料金支払済
    
            </dd>
             
            <dt
              class="col-sm-4"
            >
              
      公開ステータス
    
            </dt>
             
            <dd
              class="col-sm-8"
            >
              
      確定
    
            </dd>
          </dl>
           
          <p
            class="h5"
          >
            
    会場予約
  
          </p>
           
          <div
            data-v-5f7284b9=""
          >
            <!---->
            <div
              class="position-relative table-responsive"
              data-v-5f7284b9=""
            >
              <table
                class="table mb-0"
                data-v-5f7284b9=""
              >
                <thead
                  data-v-5f7284b9=""
                >
                  <!---->
                  <!---->
                </thead>
                <tbody
                  class="position-relative"
                  data-v-5f7284b9=""
                >
                  <tr
                    data-v-5f7284b9=""
                  >
                    <td
                      class=""
                      data-v-5f7284b9=""
                    >
                       10:00 
                    </td>
                    <td
                      class=""
                      data-v-5f7284b9=""
                    >
                       13:00 
                    </td>
                    <td
                      data-v-5f7284b9=""
                    >
                      
        会計精算済
      
                    </td>
                    <td
                      class="my-action"
                      data-v-5f7284b9=""
                    >
                      <a
                        class="btn btn-outline-primary"
                        data-v-5f7284b9=""
                        href="/reservation/reservation/split/8"
                        target="_self"
                      >
                        
          分割
        
                      </a>
                    </td>
                  </tr>
                  <!---->
                  <tr
                    data-v-5f7284b9=""
                  >
                    <td
                      class=""
                      data-v-5f7284b9=""
                    >
                       13:00 
                    </td>
                    <td
                      class=""
                      data-v-5f7284b9=""
                    >
                       16:00 
                    </td>
                    <td
                      data-v-5f7284b9=""
                    >
                      
        会計精算済
      
                    </td>
                    <td
                      class="my-action"
                      data-v-5f7284b9=""
                    >
                      <a
                        class="btn btn-outline-primary"
                        data-v-5f7284b9=""
                        href="/reservation/reservation/split/7"
                        target="_self"
                      >
                        
          分割
        
                      </a>
                    </td>
                  </tr>
                  <!---->
                  <!---->
                </tbody>
                <!---->
              </table>
              <!---->
            </div>
            <!---->
          </div>
        </div>
        <footer
          class="modal-footer"
        >
          <a
            class="btn btn-outline-primary"
            href="/reservation/schedule/edit/51"
            target="_self"
          >
            
        編集
      
          </a>
        </footer>
      </div>
    </div>
  </div>
  <div
    class="modal-backdrop fade"
  />
</div>
`;

exports[`ListModalDetails write権限：なし マウント 表示 フッター表示 1`] = `
<footer
  class="modal-footer"
>
  
         
    
</footer>
`;

exports[`ListModalDetails write権限：なし マウント 表示 予約一覧アクション表示 1`] = `
<td
  class="my-action"
  data-v-5f7284b9=""
>
  <!---->
</td>
`;
